src/haz3lcore/dynamics/Elaborator.re

```diff
@@ -241,6 +242,47 @@ let rec dhexp_of_uexp =
         let* c_fn = dhexp_of_uexp(m, fn);
         let+ c_arg = dhexp_of_uexp(m, arg);
         DHExp.Ap(c_fn, c_arg);
+      | DeferredAp(fn, args) =>
+        let* c_fn = dhexp_of_uexp(m, fn);
+        switch (err_status) {
+        | InHole(ErroneousPartialAp(Meaningless)) => Some(c_fn)
+        | InHole(ErroneousPartialAp(ArityMismatch(_))) =>
+          Some(DHExp.InvalidText(id, 0, "<inv partial ap>"))
+        | _ =>
+          let mk_tuple = (ctor, xs) =>
+            List.length(xs) == 1 ? List.hd(xs) : ctor(xs);
+          let* ty_fn = fixed_exp_typ(m, fn);
+          let (ty_arg, ty_ret) = Typ.matched_arrow(ty_fn);
+          let ty_ins = Typ.matched_args(List.length(args), ty_arg);
+          /* Substitute all deferrals for new variables */
+          let+ (pats, c_args, ty_args) =
+            List.combine(args, ty_ins)
+            |> List.fold_left(
+                 (acc, (e: Term.UExp.t, ty)) => {
+                   let* (pats, c_args, ty_args) = acc;
+                   if (Term.UExp.is_deferral(e)) {
+                     // Internal variable name for deferrals
+                     let name =
+                       "~deferred" ++ string_of_int(List.length(pats));
+                     Some((
+                       pats @ [DHPat.Var(name)],
+                       c_args @ [DHExp.BoundVar(name)],
+                       ty_args @ [ty],
+                     ));
+                   } else {
+                     let+ c_arg = dhexp_of_uexp(m, e);
+                     (pats, c_args @ [c_arg], ty_args);
+                   };
+                 },
+                 Some(([], [], [])),
+               );
+          let (pat, c_arg, ty_arg) = (
+            mk_tuple(x => DHPat.Tuple(x), pats),
+            mk_tuple(x => DHExp.Tuple(x), c_args),
+            mk_tuple(x => Typ.Prod(x), ty_args),
+          );
+          DHExp.Fun(pat, Arrow(ty_arg, ty_ret), Ap(c_fn, c_arg), None);
```

Comment:
This is going to require some manual casting... worth talking about with Cyrus, it's kind of complicated. Here's an example that will trigger a casting exception when run (the "?" are holes):

`let f: ?  -> ?  = fun a,b -> a+ b in
let g = f(1,_) in
g(1)`